src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs

```diff
@@ -596,124 +908,254 @@ protected Type TryGetType(string name)
                 var val = matches[i].Groups[6].Value.Trim();
                 paramMatches.Add(name, val);
             }
+
             return paramMatches;
+        }
+
+        /// <summary>
+        /// Detect all library references given the provided code
+        /// </summary>
+        /// <param name="code">Script code to search for CLR references</param>
+        /// <returns></returns>
+        private List<string> findClrReferences(string code)
+        {
+            var statements = new List<string>();
+            foreach (var line in code.Split(new[] { '\n', ';' }))
+            {
+                if (line.Contains("clr.AddReference"))
+                {
+                    statements.Add(line.Trim());
+                }
+            }
 
+            return statements;
         }
 
         /// <summary>
-        /// Find all import statements and import into scope.  If the type is already in the scope, this will be skipped.  
-        /// The ImportedTypes dictionary is 
+        /// Find all import statements and import into scope.  If the type is already in the scope, this will be skipped.
         /// </summary>
         /// <param name="code">The code to discover the import statements.</param>
         public void UpdateImportedTypes(string code)
         {
-            // look all import statements
-            var imports = FindBasicImportStatements(code)
-                .Union(FindTypeSpecificImportStatements(code))
-                .Union(FindAllTypeImportStatements(code));
-
-            // try and load modules into python scope
-            foreach (var import in imports)
+            // Detect all lib references prior to attempting to import anything
+            var refs = findClrReferences(code);
+            foreach (var statement in refs)
             {
-                if (scope.ContainsVariable(import.Key))
+                var previousTries = 0;
+                badStatements.TryGetValue(statement, out previousTries);
+                // TODO - why is this 3?  Should this be a constant?
+                if (previousTries > 3)
```

Comment:
@dimven Thanks for taking a look,  I think we are pretty close!  I will take a closer look at the `GetName` method as it may be related to some of the remaining 4 test failures.  It is also possible the tests are using old methods that aren't updated with the new logic.

Do you happen to remember the significance of this `3` above?  Possibly related to the length of these known assemblies?

```C#
private static string[] knownAssemblies = {
    "mscorlib",
    "RevitAPI",
    "RevitAPIUI",
    "ProtoGeometry"
};
```

Reply:
I also don't see any references to `GetName` in the current implementation which is why I marked it `Obsolete`, but I may be missing something

Reply:
Ah true. Didn't notice you split it in two methods.